Implementation for the wasi_clock_time/res

fadumina1 opened the pull request.

fadss committed:
Implementation for the wasi_clock_time/res

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -21,6 +21,7 @@
 #include "rights.h"
 #include "str.h"
 
+
```
could you please delete those new line changes?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -74,6 +79,21 @@ static __thread struct argv_environ_values *argv_environ;
 #endif
 
+uint64_t kFrequency = 0;
+void intialisingFrequency() {
```
where is this function being called?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -74,6 +79,21 @@ static __thread struct argv_environ_values *argv_environ;
 static __thread struct addr_pool *addr_pool;
 #endif
 
+uint64_t kFrequency = 0;
```
could you please use `sneak_case` for variables and function names to be consistent with the rest of the project?

fadumina1 replied:
What is an example of a sneak_case variable

loganek replied:
In this specific example it'd be `k_frequency` or just `frequency`. You can read more about it here: https://en.wikipedia.org/wiki/Snake_case

What you've used though is called `camelCase`: https://en.wikipedia.org/wiki/Camel_case

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -74,6 +79,21 @@ static __thread struct argv_environ_values *argv_environ;
+    kFrequency =  (uint64_t)Frequency.QuadPart;
+};
+
+uint64_t counter(){
```
same here, this function doesn't seem to be used

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,35 +343,72 @@ __wasi_errno_t
-    if (clock_getres(nclock_id, &ts) < 0)
-        return convert_errno(errno);
-    *resolution = convert_timespec(&ts);
+    if (BH_PLATFORM_WINDOWS) {
```
this is a compile-time constant, could you replace it with `#ifdef`'s instead?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,35 +343,72 @@ __wasi_errno_t
+            {
+                const uint64_t Result =
+                    (uint64_t)(((uint64_t)1000000000) / kFrequency);
+                *resolution = (uint32_t)Result;
```
Any reason for the `uint32_t` cast? `__wasi_timestamp_t` is an alias for `uint64_t`: https://github.com/bytecodealliance/wasm-micro-runtime/blob/81615564a1daa39e8afe9f44e9b3c4c2ec695d88/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,35 +343,72 @@ __wasi_errno_t
+                    ;
+                unlikely(!NT_SUCCESS_(Res));
+                {
+                    return convert_errno(errno);
```
Does `NtQueryTimerResolution` set the `errno` value? Could you point to a resource that defines it?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,35 +343,72 @@ __wasi_errno_t
+wasmtime_ssp_clock_time_get(wasm_exec_env_t exec_env, wasi_clockid_t clock_id,
+                            wasi_timestamp_t *resolution)
+{ 
+    if (BH_PLATFORM_WINDOWS) {
```
same here, we could change it to `#ifdef`

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -2954,7 +3012,7 @@ void wasmtime_ssp_proc_exit(
 wasmtime_ssp_proc_raise(__wasi_signal_t sig)
 {
-    static const int signals[] = {
+    static const int signals[50] = {
```
this change is not relevant to the PR (and is probably not valid either); could you revert it?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,35 +343,72 @@ __wasi_errno_t
+
+        if(!validate_native_addr(resolution, sizeof(wasi_timestamp_t)))
+            return (wasi_errno_t)-1;
+
```
looks like we're missing an implementation here?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -33,6 +34,7 @@ min(size_t a, size_t b)
     return a > b ? b : a;
 }
 
+
```
please remove the extra empty line

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -1554,6 +1708,7 @@ wasmtime_ssp_fd_allocate(
     struct stat sb;
     int ret = fstat(fd_number(fo), &sb);
     off_t newsize = (off_t)(offset + len);
+
```
please remove the empty line

loganek replied:
ping

loganek replied:
can you please remove this whitespace change?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -3035,7 +3190,7 @@ wasmtime_ssp_proc_raise(__wasi_signal_t sig)
 #endif
 #undef X
-    };
+};
```
please revert this change

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -20,6 +20,7 @@
 #include "refcount.h"
 #include "rights.h"
 #include "str.h"
+#include "ntdll.h" 
```
ntdll.h header is not available on non-windows platforms, so I think it should be included conditionally

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -74,6 +76,43 @@ static __thread struct argv_environ_values *argv_environ;
+
+
+static uint64
+calculate_monotonic_clock_frequency()()
```
why are there is the additional pair of `()`  at the end?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
                            __wasi_timestamp_t *resolution)
-{
+
 #ifdef BH_PLATFORM_WINDOWS
```
I see the implementation for posix and windows platforms is quite different. I'd suggest we:

1. add new functions to the core/shared/platform/common/posix/posix_time.c: `os_clock_res_get` and `os_clock_time_get()` and move the original implementation from this file to those functions
2. `os_clock_res_get` and `os_clock_time_get()`  to core/shared/platform/windows/win_time.c and move your windows-specific implementation there
3. in here, call the os_clock_res_get and os_clock_time_get functions instead

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+            case __WASI_CLOCK_MONOTONIC:
+            {
+                const uint64_t Result =
+                    (uint64_t)(((uint64_t)1000000000) / calculateFrequency());
```
where's the calculateFrequency() defined?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+                const uint64_t Result =
+                    (uint64_t)(((uint64_t)1000000000) / calculateFrequency());
+                *resolution = (uint64_t)Result;
+                __wasi_errno_t error = 0;
+                return error;
```
```suggestion
                return __WASI_ESUCCESS;
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+            {
+                const uint64_t Result =
+                    (uint64_t)(((uint64_t)1000000000) / calculateFrequency());
+                *resolution = (uint64_t)Result;
```
```suggestion
                *resolution = (__wasi_timestamp_t)result;
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+        switch (clock_id) {
+            case __WASI_CLOCK_MONOTONIC:
+            {
+                const uint64_t Result =
```
```suggestion
                const __wasi_timestamp_t result =
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+                 }
+                 time = Nanoseconds;
+                 
+                 return -1;
```
is this error code correct? Shouldn't we return success here?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+                              + (Fractions * NANOSECONDS_PER_SECOND)
+                                        / calculate_monotonic_clock_frequency();
+                 }
+                 time = Nanoseconds;
```
is that correct? Shouldn't we assign the nanoseconds to the time's value, i.e. `*time = Nanoseconds;`?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+                    time = from_file_time_to_wasi_timestamp(KernalTime)
+                           + from_file_time_to_wasi_timestamp(UserTime);
+
+                    return -1;
```
same here, why don't we return success here?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+
+                     time = from_file_time_to_wasi_timestamp(KernalTime)
+                           + from_file_time_to_wasi_timestamp(UserTime);
+                    return -1;
```
same here, why don't we return success here?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -322,36 +363,149 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+                    FILETIME KernalTime;
+                    FILETIME UserTime;
+
+                    if (unlikely(!GetProcessTimes(GetCurrentProcess(),
```
in what case the function can fail? Based on that we should find the appropriate error code.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -74,6 +76,43 @@ static __thread struct argv_environ_values *argv_environ;
+{
+    __wasi_timestamp_t *attempt;
+        
+    static const uint64 NTToUnixEpoch =
```
It'd be good to explain (or add a reference to a resource) where does the value come from.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +328,36 @@ __wasi_errno_t
-    return __WASI_ENOSYS;
+
+    #ifdef BH_PLATFORM_WINDOWS
+    os_win_clock_time_get(clock_id, resolution);
```
we should probably name the function `os_clock_res_get` so we don't need those ifdefs here.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +328,36 @@ __wasi_errno_t
-    *time = convert_timespec(&ts);
+{ 
+    #ifdef BH_PLATFORM_WINDOWS 
+        os_win_clock_time_get(clock_id, precision, time);  
```
same here, we probably could name this function `os_clock_time_get` and avoid ifdefs.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +328,36 @@ __wasi_errno_t
+    }
     return 0;
 #endif
+    return true ;
```
`true` resolves to 1, which is error. We need to return success here.

loganek commented on product-mini/platforms/windows/CMakeLists.txt:
```diff
@@ -151,6 +152,11 @@ set_target_properties (libiwasm PROPERTIES OUTPUT_NAME libiwasm)
   target_link_libraries (libiwasm ws2_32)
 endif ()
+
+if (WIN32)
```
is it possible to merge MING with WIN32? Are both being used when you build on windows?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/ntdll:
is this submodule actually needed? Or could we remove it?

loganek commented on product-mini/platforms/windows/CMakeLists.txt:
```diff
@@ -151,6 +152,15 @@ set_target_properties (libiwasm PROPERTIES OUTPUT_NAME libiwasm)
 endif ()
+
+if (WIN32)
+  target_link_libraries(libiwasm ntdll)
```
I think ntdll should only be linked if WASI libc is enabled, otherwise there's no need for that.

fadumina1 replied:
Should a check be added if WASI libc is enables

loganek replied:
I think the check should be (pseudocode):
if (WASI_LIBC is enabled AND MSVC is enabled)

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +325,47 @@ __wasi_errno_t
 {
-#ifdef BH_PLATFORM_WINDOWS
-    return __WASI_ENOSYS;
+
```
As suggested earlier, I think we could have a `os_clock_res_get` function defined in `win_time.c` and `posix_time.c` so this function's implementation just calls that `os_clock_res_get` method, and depending on build configuration, the correct code will be executed.  That way we can avoid lots of ifdefs and keep different implementations separate.

fadumina1 replied:
I tried to do this part but functions like convert_timespec and convert_errno are in posix.c, Should I copy and paste these implementations, wasn't sure if that was appropriate to do.

loganek replied:
that's a good point. I wonder if we could possibly slightly change the implementation so the convert_errno and convert_timespec is kept in the posix.c file, and the os_ functions return `struct timespec` instead?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +325,47 @@ __wasi_errno_t
-    if (clock_getres(nclock_id, &ts) < 0)
-        return convert_errno(errno);
-    *resolution = convert_timespec(&ts);
+    {
```
everything in the #else should probably be moved to the `os_clock_res_get` function in the `posix_time.c`

fadumina1 replied:
Including the implementation of convert_errno and convert_timespec (based on previous comment)

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +325,47 @@ __wasi_errno_t
-        return convert_errno(errno);
-    *time = convert_timespec(&ts);
+{ 
+    #ifdef BH_PLATFORM_WINDOWS 
```
similarly to `wasmtime_ssp_clock_res_get`, we could have `os_clock_time_get` in a platform-specific file. Please see my comment above.

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -4,6 +4,56 @@
+#define NANOSECONDS_PER_SECOND 1000000000
+
+typedef uint32_t __wasi_clockid_t;
+#define __WASI_CLOCK_REALTIME (0)
```
I'd suggest having `BH_CLOCK_*` defines (or enum) in `platform_common.h` so it's available for both posix and windows implementation.

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -4,6 +4,56 @@
+#define __WASI_CLOCK_PROCESS_CPUTIME_ID (2)
+#define __WASI_CLOCK_THREAD_CPUTIME_ID (3)
+
+typedef uint64_t __wasi_timestamp_t;
```
I think we don't need this typedef, you can use `uint64` when needed.

fadumina1 replied:
I get variable "__wasi_timestamp_t" is not a type name when removed, this error comes from declaring from_file_time_to_wasi_timestamp(FILETIME Filetime) as a __wasi_timestamp_t function. In that case should I keep the tyedef

loganek replied:
no, I think we can remove the typedef from this file, and replace `__wasi_timestamp_t` usage with the `uint64`

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -4,6 +4,56 @@
+
+    static const uint64 NTToUnixEpoch = 134774 * 86400 * NANOSECONDS_PER_SECOND;
+
+    ULARGE_INTEGER Temp = { .LowPart = Filetime.dwLowDateTime,
```
please use `camel_case` for variable names.

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
 }
+
+uint64
+os_win_clock_res_get(__wasi_clockid_t clock_id, __wasi_timestamp_t *resolution)
```
```suggestion
os_clock_res_get(bh_clock_id_t clock_id, uint64 *resolution)
```

And in addition to that, I'd define enum in the `platform_common.h` with all the clock IDs, something like:
```c
typedef enum {
  BH_CLOCK_REALTIME,
 ....
} bh_clock_id_t;
```

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
+    switch (clock_id) {
+        case __WASI_CLOCK_MONOTONIC:
+        {
+             const __wasi_timestamp_t result =
+                (uint64_t)(((uint64_t)1000000000)
+                           / current_value_of_peformance_counter());
```
const uint64 result = (uint64)NANOSECONDS_PER_SECOND / current_value_of_performance_counter();

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
+    switch (clock_id) {
+        case __WASI_CLOCK_MONOTONIC:
+        {
+             const __wasi_timestamp_t result =
+                (uint64_t)(((uint64_t)1000000000)
+                           / current_value_of_peformance_counter());
```
```suggestion
            const uint64 result = (uint64)NANOSECONDS_PER_SECOND / current_value_of_performance_counter();
```

loganek replied:
But I wonder if this implementation is actually correct? This function is supposed to return return a resolution of the clock, so relying on the performance counter doesn't sound right, as it will change over time (whereas the resolution should be constant).

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
+             const __wasi_timestamp_t result =
+                (uint64_t)(((uint64_t)1000000000)
+                           / current_value_of_peformance_counter());
+            *resolution = (__wasi_timestamp_t)result;
```
```suggestion
            *resolution = result;
```

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
+                (uint64_t)(((uint64_t)1000000000)
+                           / current_value_of_peformance_counter());
+            *resolution = (__wasi_timestamp_t)result;
+            __wasi_errno_t error = 0;
```
```suggestion
```
it's not used

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
     return ((uint64)ts.tv_sec) * 1000 * 1000 + ((uint64)ts.tv_nsec) / 1000;
 }
+
+uint64
```
we could probably change a return type of this function (and the clock_time_get one) to return `int`, and return either `BHT_OK` on success or `BHT_ERROR` on failure. See `posix_socket.c` for examples.

I don't think we'll need granular error value for those functions.

loganek replied:
The return type can be an `int`

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +68,126 @@ os_time_get_boot_microsecond()
+            PULONG CurrentTime;
+            NTSTATUS
+            NTAPI Res;
+            if (Res = NtQueryTimerResolution(&MaximumTime, &MinimumTime,
+                                             &CurrentTime))
+                ;
```
This doesn't seem to be correct; looks like whatever the result is, we'll ignore it, and always jump to line 98

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h:
```diff
@@ -654,17 +654,6 @@ __wasi_errno_t wasmtime_ssp_args_sizes_get(
     size_t *argv_buf_size
 ) WASMTIME_SSP_SYSCALL_NAME(args_sizes_get) WARN_UNUSED;
 
-__wasi_errno_t wasmtime_ssp_clock_res_get(
```
I think those should stay here; any reason you deleted those declarations?

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -4,6 +4,7 @@
  */
 
 #include "platform_api_vmcore.h"
+#include <winternl.h>
```
I wouldn't expect windows-specific includes in the posix implementation - do you need this include for anything?

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +16,142 @@ os_time_get_boot_microsecond()
+uint64
+os_clock_res_get(bh_clock_id_t clock_id, uint64 *resolution)
+{
+    #ifdef BH_PLATFORM_WINDOWS 
```
I think we should only keep the posix implementation in this file; windows implementation should be in the `win_time.c`

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +16,142 @@ os_time_get_boot_microsecond()
+uint64
+os_clock_time_get(bh_clock_id_t clock_id, uint64 precision, uint64 *time)
+{
+    #ifdef BH_PLATFORM_WINDOWS 
```
Same here; windows implementation should be moved to `win_time.c`

loganek commented on core/shared/platform/include/platform_common.h:
```diff
@@ -37,6 +38,13 @@ extern "C" {
 #define BH_TIME_T_MAX LONG_MAX
 #endif
 
+typedef enum{
+#define BH_CLOCK_REALTIME (0)
+#define BH_CLOCK_MONOTONIC (1)
+#define BH_CLOCK_PROCESS_CPUTIME_ID (2)
+#define BH_CLOCK_THREAD_CPUTIME_ID (3)
+} bh_clock_id_t;
```
```suggestion
typedef enum{
    BH_CLOCK_ID_REALTIME,
    BH_CLOCK_ID_MONOTONIC,
    BH_CLOCK_ID_PROCESS_CPUTIME_ID,
    BH_CLOCK_ID_THREAD_CPUTIME_ID,
} bh_clock_id_t;
```

fadumina1 replied:
Would I need to define them outside of the enum block

fadumina1 replied:
Removing the e.g #define from (e.g) BH_CLOCK_ID_REALTIME, says it's not defined now in win_time.c

loganek replied:
could you share the change you've made?

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +52,127 @@ os_time_get_boot_microsecond()
+        case BH_CLOCK_PROCESS_CPUTIME_ID:
+        case BH_CLOCK_THREAD_CPUTIME_ID:
+        {
+            PULONG MaximumTime;
+            PULONG MinimumTime;
+            PULONG CurrentTime;
```
```suggestion
            PULONG maximumTime;
            PULONG minimumTime;
            PULONG currentTime;
```

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +52,127 @@ os_time_get_boot_microsecond()
+
+        case BH_CLOCK_REALTIME:
+        {
+            FILETIME SysNow;
```
```suggestion
            FILETIME sysNow;
```

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +16,142 @@ os_time_get_boot_microsecond()
+            return __WASI_EINVAL;
+        struct timespec ts;
+        if (clock_getres(nclock_id, &ts) < 0)
+            return convert_errno(errno);
```
I'd suggest returning `BHT_ERROR` here, and then do `convert_errno()` in the `posix.c` file - I don't think `convert_errno` is available in this file.

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +16,142 @@ os_time_get_boot_microsecond()
+        case BH_CLOCK_REALTIME:
+        case BH_CLOCK_PROCESS_CPUTIME_ID:
+        case BH_CLOCK_THREAD_CPUTIME_ID:
+        {
```
looks like we never update `resolution` variable for those clock IDs?

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +52,127 @@ os_time_get_boot_microsecond()
+            PULONG MinimumTime;
+            PULONG CurrentTime;
+            NTSTATUS
+            NTAPI Res;
```
`NTAPI` can be removed here as it only defines calling convention.

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +52,127 @@ os_time_get_boot_microsecond()
+            NTAPI Res;
+            if (Res = NtQueryTimerResolution(&MaximumTime, &MinimumTime,
+                                             &CurrentTime)) {
+                !NT_SUCCESS(Res);
```
Not sure if I understand this line; could you clarify?

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +52,127 @@ os_time_get_boot_microsecond()
+            if (NANOSECONDS_PER_SECOND
+                           % calculate_monotonic_clock_frequency()
+                       == 0) {
+                Nanoseconds = Counter
```
Is this calculation correct? Looks like we only use clock frequency to calculate the time?

loganek commented on core/shared/platform/windows/win_time.c:
```diff
@@ -18,3 +52,127 @@ os_time_get_boot_microsecond()
+        }
+
+        case BH_CLOCK_MONOTONIC:
+        {
```
Where's the closing one for this `{`? I'd expect it to be in line 134 but it's not there.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +325,17 @@ __wasi_errno_t
-    *resolution = convert_timespec(&ts);
-    return 0;
-#endif
+    os_clock_res_get(clock_id, *resolution);
+    return convert_errno(errno);
```
```suggestion
    if (os_clock_res_get(clock_id, *resolution) != BHT_OK)
        return convert_errno(errno);
    return __WASI_SUCCESS;
```
that way we only convert errno if there was an error, otherwise, we'll return success code. And I think we should do the same for `wasmtime_ssp_clock_time_get`

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +53,28 @@ os_time_get_boot_microsecond()
+    if (clock_getres(clock_id, &ts) < 0)
+        return BHT_ERROR;
+    *resolution = convert_timespec(&ts);
+
```
we should return `BHT_OK` on success.

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +53,28 @@ os_time_get_boot_microsecond()
     return ((uint64)ts.tv_sec) * 1000 * 1000 + ((uint64)ts.tv_nsec) / 1000;
 }
+
+uint64
```
```suggestion
int
```

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +53,28 @@ os_time_get_boot_microsecond()
+uint64
+os_clock_res_get(bh_clock_id_t clock_id, uint64 *resolution)
+{
+    bh_clock_id_t nclock_id;
```
```suggestion
    clockid_t nclock_id;
```
because `convert_clockid` returns posix's `clockid_t` value

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +53,28 @@ os_time_get_boot_microsecond()
+
+}
+
+uint64
```
```suggestion
int
```

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +53,28 @@ os_time_get_boot_microsecond()
+uint64
+os_clock_time_get(bh_clock_id_t clock_id, uint64 precision, uint64 *time)
+{
+    bh_clock_id_t nclock_id;
```
```suggestion
    clockid_t nclock_id;
```

see comment above.

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +53,28 @@ os_time_get_boot_microsecond()
+    if (clock_gettime(nclock_id, &ts) < 0)
+        return BHT_ERROR;
+        *time = convert_timespec(&ts);
+}
```
similar here, this function should return `BHT_OK` on success.

loganek commented on core/shared/platform/include/platform_api_vmcore.h:
```diff
@@ -65,6 +65,12 @@ os_vprintf(const char *format, va_list ap);
 uint64
 os_time_get_boot_microsecond(void);
+/* Get current clock time*/
+uint64
```
```suggestion
int
```

loganek commented on core/shared/platform/include/platform_api_vmcore.h:
```diff
@@ -65,6 +65,12 @@ os_vprintf(const char *format, va_list ap);
+uint64
+os_clock_time_get(bh_clock_id_t clock_id, uint64 precision, uint64 *time);
+/* Get current resolution time*/
+uint64
```
```suggestion
int
```

loganek commented on core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:
```diff
@@ -192,7 +192,7 @@ wasi_clock_res_get(wasm_exec_env_t exec_env,
         return (wasi_errno_t)-1;
 
-    return wasmtime_ssp_clock_res_get(clock_id, resolution);
+    return os_clock_res_get(clock_id, resolution);
```
I think this should stay as it was, i.e.
```suggestion
    return wasmtime_ssp_clock_res_get(clock_id, resolution);
```

loganek commented on core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c:
```diff
@@ -206,7 +206,7 @@ wasi_clock_time_get(wasm_exec_env_t exec_env,
         return (wasi_errno_t)-1;
 
-    return wasmtime_ssp_clock_time_get(clock_id, precision, time);
+    return os_clock_time_get(clock_id, precision, time);
```
same here, please revert this change
```suggestion
    return wasmtime_ssp_clock_time_get(clock_id, precision, time);
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -178,6 +178,8 @@ convert_errno(int error)
 }
 
+
+
```
```suggestion
```

remove unused empty lines

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -3035,7 +3020,7 @@ wasmtime_ssp_proc_raise(__wasi_signal_t sig)
 #endif
 #undef X
-    };
+};
```
```suggestion
    };
```

loganek commented on core/shared/platform/include/platform_api_vmcore.h:
```diff
@@ -65,6 +65,12 @@ os_vprintf(const char *format, va_list ap);
  */
 uint64
 os_time_get_boot_microsecond(void);
+/* Get current clock time*/
```
because the function doesn't return the clock time, it's worth documenting that it returns BHT_OK on success and BHT_ERROR on failure. The same for res_get function.

loganek replied:
```suggestion

/* Get current clock time*/
```
also, please add extra empty line between previous function and this comment.

loganek commented on core/shared/platform/include/platform_common.h:
```diff
@@ -13,6 +13,7 @@ extern "C" {
 #include "platform_internal.h"
 #include "../../../config.h"
 
+
```
```suggestion
```

loganek commented on core/shared/platform/windows/platform_internal.h:
```diff
@@ -31,6 +31,7 @@
 #include <basetsd.h>
 #include <signal.h>
 
+
```
```suggestion
```

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -4,6 +4,43 @@
+#include <time.h>
+
+uint64
+convert_timespec(const struct timespec *ts)
```
to avoid duplicating the implementation from the posix.c file, maybe we can create a utility file in the `core/shared/utils` (let's call it bh_time.c / bh_time.h) so it's available both in posix_time.c and posix.c file.

Alternatively, we can update the interface for the os_clock_* functions so they return timespec instead of uint64.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -323,37 +325,19 @@ __wasi_errno_t
-    if (clock_gettime(nclock_id, &ts) < 0)
+                            __wasi_timestamp_t *time){
+    
+    if(os_clock_time_get(clock_id,precision,*time) !=BHT_OK);
```
I think you don't mean to have the semicollon at the end
```suggestion
    if(os_clock_time_get(clock_id,precision,*time) !=BHT_OK)
```

loganek commented on CMakeLists.txt:
```diff
@@ -131,7 +131,7 @@ endif ()
 
 # STATIC LIBRARY
-add_library(iwasm_static STATIC ${WAMR_RUNTIME_LIB_SOURCE})
+add_library(iwasm_static STATIC ${WAMR_RUNTIME_LIB_SOURCE} "core/shared/utils/bh_time.c" "core/shared/utils/bh_time.h")
```
this should not be needed, it's already included through `shared_utils.cmake`

loganek commented on CMakeLists.txt:
```diff
@@ -142,7 +142,7 @@ endif ()
 
 # SHARED LIBRARY
-add_library (iwasm_shared SHARED ${WAMR_RUNTIME_LIB_SOURCE})
+add_library (iwasm_shared SHARED ${WAMR_RUNTIME_LIB_SOURCE} "core/shared/utils/bh_time.c" "core/shared/utils/bh_time.h")
```
same here as above; `shared_utils.cmake` already includes this file

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +16,31 @@ os_time_get_boot_microsecond()
+        return BHT_ERROR;
+    struct timespec ts;
+    if (clock_gettime(nclock_id, &ts) < 0)
+        return BHT_OK;
```
Shouldn't that be an error?

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -15,3 +16,31 @@ os_time_get_boot_microsecond()
+        return BHT_OK;
+    *time = convert_timespec(&ts);
+
+    return 0;
```
```suggestion
    return BHT_OK;
```

loganek commented on core/shared/utils/bh_time.c:
```diff
@@ -0,0 +1,42 @@
+            return true;
+#endif
+        default:
+            return false;
```
We should probably set errno here as well.
```suggestion
            errno = EINVAL;
            return false;
```

loganek commented on core/shared/utils/bh_time.c:
```diff
@@ -0,0 +1,42 @@
+}
+
+static bool
+convert_clockid(bh_clock_id_t in, _clockid_t *out)
```
This function is posix-specific so shouldn't be in the common code. I'd suggest we move it to posix_time.c

loganek replied:
Also, this function should use `clockid_t` type:
```suggestion
convert_clockid(bh_clock_id_t in, clockid_t *out)
```

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,33 @@
+extern "C" {
+#endif
+
+#define CLOCK_REALTIME 0
+#define CLOCK_MONOTONIC 1
+#define CLOCK_PROCESS_CPUTIME_ID 2
+#define CLOCK_THREAD_CPUTIME_ID 3
```
this used to be `bh_clock_id_t` enum; any reason for this change?

loganek commented on product-mini/platforms/windows/CMakeLists.txt:
```diff
@@ -137,7 +137,8 @@ add_executable (iwasm main.c ${UNCOMMON_SHARED_SOURCE})
 
-target_link_libraries (iwasm vmlib ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS})
+target_link_libraries (iwasm vmlib  ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS} )
+
```
```suggestion
```

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,33 @@
+#define CLOCK_PROCESS_CPUTIME_ID 2
+#define CLOCK_THREAD_CPUTIME_ID 3
+
+typedef uint32 _clockid_t;
```
This should not be needed.
```suggestion
```

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,33 @@
+uint64
+convert_timespec(const struct timespec *ts);
+
+static bool
+convert_clockid(bh_clock_id_t in, _clockid_t *out);
```
This function should be implemented in posix_time.c and removed from this header.
```suggestion
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,43 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
     }
 }
 
+bh_clock_id_t
+convert_clock_to_bh_clock(__wasi_clockid_t clock_id)
```
I'd suggest we turn this function into the following:
```suggestion
bool
convert_clock_to_bh_clock(__wasi_clockid_t in, bh_clock_id_t *out)
```
That way we clearly indicate when there's an error. The function should return true on success, and false on failure.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,43 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
-        return __WASI_EINVAL;
-    struct timespec ts;
-    if (clock_getres(nclock_id, &ts) < 0)
+    convert_clock_to_bh_clock(clock_id);
```
shouldn't we use the result of this function?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -3834,7 +3797,8 @@ argv_environ_init(struct argv_environ_values *argv_environ, char *argv_buf,
 void
 argv_environ_destroy(struct argv_environ_values *argv_environ)
-{}
+{
+}
```
please revert whitespace changes.

loganek commented on core/shared/platform/include/platform_api_vmcore.h:
```diff
@@ -66,6 +66,14 @@ os_vprintf(const char *format, va_list ap);
 uint64
 os_time_get_boot_microsecond(void);
 
+/*y*/
```
Please update the documentation

loganek commented on core/shared/platform/include/platform_api_vmcore.h:
```diff
@@ -66,6 +66,14 @@ os_vprintf(const char *format, va_list ap);
+int
+os_clock_res_get(bh_clock_id_t clock_id, uint64 *resolution);
+
+/*t*/
```
please update the documentation

loganek commented on core/shared/platform/include/platform_api_vmcore.h:
```diff
@@ -66,6 +66,14 @@ os_vprintf(const char *format, va_list ap);
+
+/*t*/
+int
+os_clock_time_get(bh_clock_id_t clock_id, uint64 precision, uint64 *time);
```
I think those two functions probably can be moved to platform_common.h

loganek commented on core/shared/platform/linux-sgx/shared_platform.cmake:
```diff
@@ -8,6 +8,7 @@ add_definitions(-DBH_PLATFORM_LINUX_SGX)
 include_directories(${PLATFORM_SHARED_DIR})
 include_directories(${PLATFORM_SHARED_DIR}/../include)
 
+
```
whitespace change, please remove

loganek commented on core/shared/utils/bh_time.c:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
```
```suggestion
* Copyright (C) 2023 Amazon Inc.  All rights reserved.
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
```
```suggestion
* Copyright (C) 2023 Amazon Inc.  All rights reserved.
```

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,20 @@
+
+#ifndef _BH_TIME_
+#define _BH_TIME_
+#include "bh_platform.h"
```
why do we have to include this file?

fadumina1 replied:
In the bh_platform.h is quite similar to platform_common.h and connects the header files to use typedef like unit64 which bh_time.c needs. All the files have bh_platform.h in their header files.

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,20 @@
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef _BH_TIME_
+#define _BH_TIME_
```
```suggestion
#ifndef _BH_TIME_H
#define _BH_TIME_H
```

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
```
```suggestion
* Copyright (C) 2023 Amazon Inc.  All rights reserved.
```

loganek commented on product-mini/platforms/alios-things/aos.mk:
```diff
@@ -102,6 +102,7 @@ $(NAME)_SOURCES := ${SHARED_ROOT}/platform/alios/alios_platform.c \
                    ${SHARED_ROOT}/utils/bh_hashmap.c \
                    ${SHARED_ROOT}/utils/bh_list.c \
                    ${SHARED_ROOT}/utils/bh_log.c \
+		           ${SHARED_ROOT}/utils/bh_time.c \
```
Please align the line with the others

loganek commented on product-mini/platforms/nuttx/wamr.mk:
```diff
@@ -348,6 +348,7 @@ endif
 CSRCS += nuttx_platform.c \
          posix_thread.c \
          posix_time.c \
+	     posix_clock.c \
```
please align with other files

loganek replied:
This is still not aligned.

loganek commented on product-mini/platforms/nuttx/wamr.mk:
```diff
@@ -357,6 +358,7 @@ CSRCS += nuttx_platform.c \
          bh_hashmap.c \
          bh_list.c \
          bh_log.c \
+	     bh_time.c \
```
please align with other files

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+uint64
+current_value_of_peformance_counter()
+{
+    LARGE_INTEGER Counter;
```
```suggestion
    LARGE_INTEGER counter;
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+static uint64
+from_file_time_to_wasi_timestamp(FILETIME Filetime)
+{
+
```
remove empty line

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+// The implementation below dervies from the following source:
+// https://github.com/WasmEdge/WasmEdge/blob/b70f48c42922ce5ee7730054b6ac0b1615176285/lib/host/wasi/win.h#L210
+static uint64
+from_file_time_to_wasi_timestamp(FILETIME Filetime)
```
```suggestion
from_file_time_to_wasi_timestamp(FILETIME filetime)
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+os_clock_time_get(bh_clock_id_t clock_id, uint64 precision, uint64 *time)
+{
+    switch (clock_id) {
+
```
remove empty line

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+
+#endif
+            time = from_file_time_to_wasi_timestamp(sysNow);
+            return 0;
```
```suggestion
            return BHT_OK;
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+        case BH_CLOCK_ID_MONOTONIC:
+        {
+            uint64_t Nanoseconds;
+            const auto Counter = calculate_monotonic_clock_frequency();
```
```suggestion
            uint64 counter = calculate_monotonic_clock_frequency();
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+
+        case BH_CLOCK_ID_MONOTONIC:
+        {
+            uint64_t Nanoseconds;
```
This probably can be removed
```suggestion
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+            const auto Counter = calculate_monotonic_clock_frequency();
+            if (NANOSECONDS_PER_SECOND % calculate_monotonic_clock_frequency()
+                == 0) {
+                Nanoseconds = Counter
```
```suggestion
                *time = counter
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+                    Counter / calculate_monotonic_clock_frequency();
+                const auto Fractions =
+                    Counter % calculate_monotonic_clock_frequency();
+                Nanoseconds = Seconds * NANOSECONDS_PER_SECOND
```
```suggestion
                *time = seconds * NANOSECONDS_PER_SECOND
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+                              + (Fractions * NANOSECONDS_PER_SECOND)
+                                    / calculate_monotonic_clock_frequency();
+            }
+            *time = Nanoseconds;
```
```suggestion
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+                                 / calculate_monotonic_clock_frequency());
+            }
+            else {
+                const auto Seconds =
```
please don't use auto - it's c++ feature and this file is C. Even though it will work in msvc (because it compiles it with C++ compiler anyway), better to not introduce inconsistency.

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+
+            case BH_CLOCK_ID_PROCESS_CPUTIME_ID:
+            {
+                FILETIME CreationTime;
```
please use `sneak_case` for variable names, e.g. `creation_time`

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+
+            return BHT_OK;
+
+            case BH_CLOCK_ID_PROCESS_CPUTIME_ID:
```
isn't the implementation of BH_CLOCK_ID_PROCESS_CPUTIME_ID and BH_CLOCK_ID_THREAD_CPUTIME_ID identical?

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,153 @@
+            return 0;
+        }
+
+        case BH_CLOCK_ID_MONOTONIC:
```
is the implementation of this clock correct? Looks like we only use frequency to calculate time, which seems a bit odd.

loganek replied:
I have looked at the wasmedge implementation for you; I think the counter value you read is frequency, and not the actual counter. Can you please update the code.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -280,9 +280,9 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
 }
 
 bh_clock_id_t
```
```suggestion
bool
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
     }
 }
 
+bh_clock_id_t
```
```suggestion
bool
```
the bh clock id should be returned through the `out` variable, e.g. `*out = BH_CLOCK_ID_MONOTONIC;`

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+        case __WASI_CLOCK_THREAD_CPUTIME_ID:
+            return BH_CLOCK_ID_THREAD_CPUTIME_ID;
+        default:
+            return convert_errno(errno);
```
```suggestion
            return false;
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
-    struct timespec ts;
-    if (clock_getres(nclock_id, &ts) < 0)
+    bh_clock_id_t bh_clockdid;
+    if (convert_clock_to_bh_clock(!clock_id, &bh_clockdid))
```
Why do we negate `clock_id` here?

loganek replied:
Also, this should probably be 
```suggestion
    if (!convert_clock_to_bh_clock(!clock_id, &bh_clockdid))
```
if we update `convert_clock_to_bh_clock` to return `bool`

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
 }
 
+bh_clock_id_t
+convert_clock_to_bh_clock(__wasi_clockid_t in, bh_clock_id_t *out)
```
```suggestion
convert_wasi_clock_id_to_bh_clock_id(__wasi_clockid_t in, bh_clock_id_t *out)
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
-        return __WASI_EINVAL;
-    struct timespec ts;
-    if (clock_gettime(nclock_id, &ts) < 0)
+    bh_clock_id_t bh_clockdid;
```
Typo?
```suggestion
    bh_clock_id_t bh_clock_id;
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
-        return __WASI_EINVAL;
-    struct timespec ts;
-    if (clock_getres(nclock_id, &ts) < 0)
+    bh_clock_id_t bh_clockdid;
```
Typo?
```suggestion
    bh_clock_id_t bh_clock_id;
```

loganek commented on core/shared/platform/common/posix/posix_time.c:
```diff
@@ -4,6 +4,7 @@
  */
 
 #include "platform_api_vmcore.h"
+#include "bh_platform.h"
```
why do we need this include?

loganek commented on core/shared/utils/bh_time.h:
```diff
@@ -0,0 +1,19 @@
+#ifdef __cplusplus
+}
+#endif
+#endif /* end of _BH_TIME_ */
```
```suggestion
#endif /* end of _BH_TIME_T */
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,148 @@
+static uint64
+from_file_time_to_wasi_timestamp(FILETIME filetime)
+{
+    static const uint64 nTToUnixEpoch = 134774 * 86400 * NANOSECONDS_PER_SECOND;
```
please use `sneak_case` for variable names. Also, please review all your changes and address this comment.

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,148 @@
+        case BH_CLOCK_ID_PROCESS_CPUTIME_ID:
+        case BH_CLOCK_ID_THREAD_CPUTIME_ID:
+        {
+
```
please remove the empty line here.

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,148 @@
+int
+os_clock_res_get(bh_clock_id_t clock_id, uint64 *resolution)
+{
+    switch (clock_id) {
```
we should have default branch here that sets errno to  EINVAL and return BHT_ERROR

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,47 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
     }
 }
 
+bool
```
Please note this function returns bool, and the converted clock value should be passed through the `out`parameter. In your case, you return the converted value from the function.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,51 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
-    if (clock_getres(nclock_id, &ts) < 0)
+    bh_clock_id_t bh_clockid;
+    if (!convert_wasi_clock_id_to_bh_clock_id(clock_id, &bh_clockid))
         return convert_errno(errno);
```
```suggestion
        return __WASI_EINVAL;
```

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,51 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
-    if (clock_gettime(nclock_id, &ts) < 0)
+    bh_clock_id_t bh_clockid;
+    if (!convert_wasi_clock_id_to_bh_clock_id(clock_id, &bh_clockid))
         return convert_errno(errno);
```
```suggestion
        return __WASI_EINVAL;
```

loganek commented on core/shared/platform/common/posix/posix_clock.c:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
```
```suggestion
 * Copyright (C) 2023 Amazon Inc.  All rights reserved.
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,154 @@
+    ULARGE_INTEGER temp = { .LowPart = filetime.dwLowDateTime,
+                            .HighPart = filetime.dwHighDateTime };
+
+    auto duration = temp.QuadPart - ntto_unix_epoch;
```
As mentioned in previous revisions, please don't use `auto` keyword as it's confusing in C files.

loganek replied:
```suggestion
    return temp.QuadPart - ntto_unix_epoch;
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,154 @@
+                == 0) {
+                *time = counter
+                        * (NANOSECONDS_PER_SECOND
+                           / calculate_monotonic_clock_frequency());
```
instead of calculating the frequency multiple times in this function, you can do it once, cache it in a variable and use the variable instead.

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,154 @@
+            *time = from_file_time_to_wasi_timestamp(kernal_time)
+                    + from_file_time_to_wasi_timestamp(user_time);
+
+            return BHT_ERROR;
```
```suggestion
            return BHT_OK;
```

loganek commented on product-mini/platforms/alios-things/aos.mk:
```diff
@@ -102,6 +102,7 @@ $(NAME)_SOURCES := ${SHARED_ROOT}/platform/alios/alios_platform.c \
                    ${SHARED_ROOT}/utils/bh_hashmap.c \
                    ${SHARED_ROOT}/utils/bh_list.c \
                    ${SHARED_ROOT}/utils/bh_log.c \
+		   ${SHARED_ROOT}/utils/bh_time.c \
```
please use spaces for indentation, not tabs.

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -319,41 +279,51 @@ wasi_addr_ip_to_bh_ip_addr_buffer(__wasi_addr_ip_t *addr,
+convert_wasi_clock_id_to_bh_clock_id(__wasi_clockid_t in, bh_clock_id_t *out)
+{
+    switch (in) {
+
```
```suggestion
```
delete empty line

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,147 @@
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include "platform_api_vmcore.h"
```
```suggestion
#include "platform_common.h"
```

loganek commented on product-mini/platforms/alios-things/aos.mk:
```diff
@@ -102,6 +102,7 @@ $(NAME)_SOURCES := ${SHARED_ROOT}/platform/alios/alios_platform.c \
                    ${SHARED_ROOT}/utils/bh_hashmap.c \
                    ${SHARED_ROOT}/utils/bh_list.c \
                    ${SHARED_ROOT}/utils/bh_log.c \
+	        	   ${SHARED_ROOT}/utils/bh_time.c \
```
could you please fix the indentation?

loganek commented on product-mini/platforms/windows/CMakeLists.txt:
```diff
@@ -137,7 +137,7 @@ add_executable (iwasm main.c ${UNCOMMON_SHARED_SOURCE})
 install (TARGETS iwasm DESTINATION bin)
 
-target_link_libraries (iwasm vmlib ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS})
+target_link_libraries (iwasm vmlib  ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS} )
```
```suggestion
target_link_libraries (iwasm vmlib ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS} )
```
(formatting - extra space)

loganek commented on product-mini/platforms/windows/CMakeLists.txt:
```diff
@@ -154,3 +154,13 @@ target_link_libraries (libiwasm ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS})
   target_link_libraries (libiwasm ws2_32)
 endif ()
+
+if (WIN32)
+  target_link_libraries(libiwasm ntdll)
+endif()
+
+if (WIN32)
+  target_link_libraries(iwasm ntdll)
+endif()
```
```suggestion
if (WIN32)
  target_link_libraries(libiwasm ntdll)
  target_link_libraries(iwasm ntdll)
endif()
```
we can merge those two together

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,147 @@
+        case BH_CLOCK_ID_PROCESS_CPUTIME_ID:
+        case BH_CLOCK_ID_THREAD_CPUTIME_ID:
+        {
+            PULONG maximum_time;
+            PULONG minimum_time;
+            PULONG current_time;
```
```suggestion
            PULONG maximum_resolution;
            PULONG minimum_resolution;
            PULONG current_resolution;
```

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,147 @@
+            PULONG maximum_time;
+            PULONG minimum_time;
+            PULONG current_time;
+            NTSTATUS
+            status = NtQueryTimerResolution(&maximum_time, &minimum_time,
+                                            &current_time);
```
```suggestion
            NTSTATUS status = NtQueryTimerResolution(&maximum_time, &minimum_time,
                                            &current_time);
```
formatting

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,147 @@
+                                            &current_time);
+
+            uint64 result = (uint64)current_time * NANOSECONDS_PER_TICK;
+            *resolution = result / (uint64)NANOSECONDS_PER_SECOND;
```
why do we divide here by NANOSECONDS_PER_SECOND?

fadumina1 replied:
From win.h I understood that from the code: 'using FiletimeDuration = std::chrono::duration<
    uint64_t,
    std::ratio_multiply<std::ratio<100, 1>, std::chrono::nanoseconds::period>>'; (win.h) and in clock_win.cpp FiletimeDuration{CurrentResolution};

loganek replied:
The result value here should be nanoseconds, so if we divide that by `NANOSECONDS_PER_SECOND`, the result will be seconds. I think this should simply be: `*resolution = (uint64) current_time * NANOSECONDS_PER_TICK;`

loganek commented on core/shared/platform/common/posix/posix_clock.c:
```diff
@@ -0,0 +1,61 @@
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include "platform_api_vmcore.h"
```
this include is probably unused?

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,147 @@
+from_file_time_to_wasi_timestamp(FILETIME filetime)
+{
+    static const uint64 ntto_unix_epoch =
+        134774 * 86400 * NANOSECONDS_PER_SECOND;
```
aren't you getting integer overflow errors? All of those values are integers, so most likely you do. I'd suggest making all the numbers in the equation uint64

loganek commented on core/shared/platform/windows/win_clock.c:
```diff
@@ -0,0 +1,147 @@
+    ULARGE_INTEGER temp = { .LowPart = filetime.dwLowDateTime,
+                            .HighPart = filetime.dwHighDateTime };
+
+    return temp.QuadPart - ntto_unix_epoch;
```
I don't think this is correct - FILETIME structure represents the number of 100-nanosecond intervals, not nanoseconds. But you're substracting here a number of 100-nanosecond intervals from nanoseconds. You shouldn't substract two different values of different units. You'll need to first agree on a common unit (either way works fine) and then do the math.

see: https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-filetime

loganek commented on product-mini/platforms/windows/CMakeLists.txt:
```diff
@@ -149,8 +149,15 @@ install (TARGETS libiwasm DESTINATION lib)
 set_target_properties (libiwasm PROPERTIES OUTPUT_NAME libiwasm)
 
-target_link_libraries (libiwasm ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS})
+target_link_libraries (iwasm vmlib ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS} )
```
why did you introduce this change?

zoraaver committed:
Abstract POSIX filesystem functions (#2585)

To allow non-POSIX platforms such as Windows to support WASI libc
filesystem functionality, create a set of wrapper functions which provide a
platform-agnostic interface to interact with the host filesystem. For now,
the Windows implementation is stubbed but this will be implemented
properly in a future PR. There are no functional changes in this change,
just a reorganization of code to move any direct POSIX references out of
posix.c in the libc implementation into posix_file.c under the shared
POSIX sources.

See https://github.com/bytecodealliance/wasm-micro-runtime/issues/2495 for a
more detailed overview of the plan to port the WASI libc filesystem to Windows.

loganek commented:
Closing the PR as it ws already opened for WAMR main repo: https://github.com/bytecodealliance/wasm-micro-runtime/pull/2637
